File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/Rfc3339.cs
using System.Globalization;

namespace LinkKeys.LocalRp;

/// <summary>
/// RFC3339 timestamp parse/format helpers. Parsing accepts both the <c>Z</c> and
/// <c>+00:00</c> UTC-offset spellings (conformance vectors use the latter; this SDK's
/// own output uses the former) — both are valid RFC3339 and
/// <see cref="DateTimeOffset"/>'s round-trip/ISO parser accepts either.
/// </summary>
internal static class Rfc3339
{
    public static DateTimeOffset Parse(string field, string s)
    {
        if (DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.None, out var result))
        {
            return result;
        }

        throw new LocalRpError(LocalRpError.ErrorKind.BadTimestamp, $"{field}: could not parse '{s}' as RFC3339");
    }

    public static string Format(DateTimeOffset instant) =>
        instant.ToUniversalTime().ToString("yyyy-MM-dd\\THH:mm:ss.FFFFFFF\\Z", CultureInfo.InvariantCulture);
}

File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/Rfc3339.java
package community.catalyst.linkkeys.localrp;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeParseException;

/**
 * RFC3339 timestamp parse/format helpers. Parsing accepts both the {@code Z}
 * and {@code +00:00} UTC-offset spellings (conformance vectors use the
 * latter; this SDK's own output uses the former) &mdash; both are valid
 * RFC3339 and {@link OffsetDateTime}'s ISO offset parser accepts either.
 */
final class Rfc3339 {
    private Rfc3339() {}

    static Instant parse(String field, String s) {
        try {
            return OffsetDateTime.parse(s).toInstant();
        } catch (DateTimeParseException e) {
            throw new LocalRpError(LocalRpError.Kind.BAD_TIMESTAMP, field + ": " + e.getMessage());
        }
    }

    static String format(Instant instant) {
        return instant.atOffset(ZoneOffset.UTC).toString();
    }
}
